pip/pip-385.md

```diff
@@ -0,0 +1,398 @@
+                note right of Server: TCP read resumed
+                deactivate Server
+                Server-->>Server: read message 3
+                Client->>Server: Send(4, message4)
+            end
+        end
+    end
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Server-->>Client: SendReceipt(4,msgId4)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Java client changes
+
+* Client will now have logic to understand the `ThrottleProducer` command and take relevant action of blocking further
+  messages for the relevant topic. It will then respond back with `ThrottleProducerReceipt` command.
+    * Client will resume message sending after the specified time in the `ThrottleProducer` command's data.
+    * This interval of no messages will be noted as "being throttled"
+    * Within this duration, another `ThrottleProducer` command from server may come.
+* Producer will record new OTel metric indicating which topic was throttled and the reason.
+* In case a message fails due to timeout and there was a throttled command from server for the owning topic, client will
+  instead throw a rate limit exception instead of timeout exception.
+
+# Detailed Design
+
+## High-level Implementation Details
+
+### Broker Changes
+
+* For calls arising from `PublishRateLimiterImpl` class, add logic in `ServerCnxThrottleTracker.java` to send the
+  command to client and wait for response for the max configured duration before calling `changeAutoRead`. It checks for
+  feature availability first.
+* For calls arising from `ServerCnxThrottleTracker::changeThrottlingFlag`, we send the command async (if feature
+  supported) without worrying about response and then call `changeAutoRead`.
+* capture `AbstractTopic::getTotalPublishRateLimitCounter` per publish rate limit counter and add relevant attribute in
+  the rate limit metric
+
+### Determining the throttling duration for clients
+
+Broadly, there are 2 different categories for rate limiting - (a) quota exceeding at a topic/resource group level OR (b)
+connection level breaches leading to broker's resource crunch.
+
+In case (a), as per the current token bucket based quota tracking introduced in PIP-322, we (lazily) resume channel
+reads on the next second mark. i.e. the throttling is only applied at max for 1 second.
+In case (b), there is no clear picture on when the channel read is resumed as we wait till we have processed half of the
+pending messages, or cleared half of the memory threshold.
+
+Thus, for case (a), we will be asking the client to pause further message sending for the remainder of the second and in
+case (b), we will send `0` as the channel read is anyway paused.
+
+### Java Client Changes
+
+* We maintain a boolean and an enum field at a producer level to capture if the producer is being throttled or not and
+  the reason for current throttling.
```

Comment:
"slowing down" would mean anticipation.. and again, with producers being distributed.. its very tricky to handle anticipation about the future.

while the utilization can certainly be added in as a metric that can be consumed easily, this boolean gives a more direct answer.

in a partitioned producer, its even more useful while selecting the next partition in the round robin manner.

Reply:
> It does give a direct answer, but it's pretty much useless. In a real scenario, when the producer is producing faster than the limit, the throttling will be continuously switching on and off. 

We have discussed this internally with the potential users and we already have a few use cases which will greatly benefit from this.
Yes, the throttling will be continuously switching on and off, but in scenarios where the upstream can do a retry after a delay, sending them an early exit saves many resources (threads on both upstream and pulsar client side) .

This is mostly coming from use cases similar to a RestBus on top of Pulsar serving to many topics at once.

Reply:
> Yes, the throttling will be continuously switching on and off, but in scenarios where the upstream can do a retry after a delay, sending them an early exit saves many resources (threads on both upstream and pulsar client side) .

I don't think that the boolean value provides necessary information for this type of use case since the value could be unthrottled even in cases where the throttling is happening a lot (it just happens to be unthrottled when the value is read) and publish latency is high. A sliding window publish latency information would be more useful. Isn't the publish latency the relevant metric for the use case that you presented?

Reply:
> You are assuming that this boolean would be accessed only once in a while? The example that you provided, sure, that very instant, that single call is a true negative, but the very next call, or a few calls later - it would give the right result that throttling is happening.
> latency metrics cannot be used as a source of truth here as the latencies can be near "timeout" for N number of reason, not necessarily resulting in/caused by throttling on the topic.

my assumptions of how throttling will be changing on and off is perhaps very different from your assumption. It could happen multiple times a second, depending on the accuracy and configuration of the implementation. My opinion is simply that a boolean metric is not useful as a metric for the use case that you presented. A utilization (or probability) metric or a latency based metric would be more useful for such decisions.  I could always be wrong about this and we don't have to get stuck on this particular detail. It's possible to revisit later based on real usage experiences.